/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";

package e8;

import "real_time_message.proto";

message EnqueueMessageRequest {
    int64 user_id = 1;
    repeated RealTimeMessage messages = 2;
}

message EnqueueMessageResponse {
}


message DequeueMessageRequest {
    int64 user_id = 1;
    bool previous_message_delivered = 2;
    bool end_operation = 3;
    int32 wait_duration_secs = 4;
}

message DequeueMessageResponse {
    RealTimeMessage message = 2;
}

// A user oriented message queue.
service MessageQueueService {
    rpc EnqueueMessage (EnqueueMessageRequest) 
        returns (EnqueueMessageResponse);

    rpc DequeueMessage (stream DequeueMessageRequest) 
        returns (stream DequeueMessageResponse);
}
